javalib/src/main/resources/scala-native/process_monitor.c

```diff
@@ -0,0 +1,200 @@
+#if defined(SCALANATIVE_COMPILE_ALWAYS) ||                                     \
+    defined(__SCALANATIVE_JAVALIB_PROCESS_MONITOR)
+#if defined(__unix__) || defined(__unix) || defined(unix) ||                   \
+    (defined(__APPLE__) && defined(__MACH__))
+
+#include <pthread.h>
+#include <sys/wait.h>
+#include <sys/time.h>
+#include <semaphore.h>
+#include <errno.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <fcntl.h>
+#include <string.h>
+
+// The +1 accounts for the null char at the end of the name
+#ifdef __APPLE__
+#include <sys/posix_sem.h>
+#define SEM_MAX_LENGTH PSEMNAMLEN + 1
+#else
+#include <limits.h>
+#define SEM_MAX_LENGTH _POSIX_PATH_MAX + 1
+#endif
+
+typedef struct Monitor {
+    pthread_cond_t cond;
+    int *res;
+} Monitor;
+
+typedef struct WaitingProc {
+    int pid;
+    Monitor *monitor;
+    struct WaitingProc *next;
+} WaitingProc;
+
+typedef struct FinishedProc {
+    int pid;
+    int result;
+    struct FinishedProc *next;
+} FinishedProc;
+
+static pthread_mutex_t shared_mutex;
+static WaitingProc *waiting_procs = NULL;
+static FinishedProc *finished_procs = NULL;
```

Comment:
I changed the map with a linked list. We have linear access, now but we don't need to implement the hash map.